### PR TITLE
fix(cli): remove possible deadlock in test channel

### PR DESCRIPTION
**Pull request by mmastrac**
The stderr stream could possibly starve the other bits of the output-redirecting event loop.

### CODE REVIEW

File: cli/tools/test/channel.rs [hunk 1]
```diff
@@ -143,17 +148,23 @@ impl TestStream {
...
   /// [`UnboundedSender`] before returning.
-  async fn pipe(&mut self) {
+  fn poll_pipe(&mut self, cx: &mut std::task::Context) -> Poll<()> {
     let mut buffer = [0_u8; BUFFER_SIZE];
     let mut buf = ReadBuf::new(&mut buffer);
     let res = {
-      // No more stream, so just return.
+      // No more stream, we shouldn't hit this case.
       let Some(stream) = &mut self.read_opt else {
-        return;
```
Comment (mmastrac) [inline cli/tools/test/channel.rs:154]: This could lead to starvation

File: cli/tools/test/channel.rs [hunk 2]
```diff
@@ -253,7 +265,7 @@ impl TestEventSenderFactory {
         loop {
           tokio::select! {
             _ = test_stdout.pipe(), if test_stdout.is_alive() => {},
-            _ = test_stderr.pipe(), if test_stdout.is_alive() => {},
+            _ = test_stderr.pipe(), if test_stderr.is_alive() => {},
```
Comment (mmastrac) [inline cli/tools/test/channel.rs:268]: This typo here meant that we were polling test_stderr in a hot loop if it finished, never getting to the sync receiver

File: cli/tools/test/channel.rs [hunk 3]
```diff
@@ -377,7 +389,9 @@ impl TestEventSender {
     let mutex = parking_lot::RawMutex::INIT;
     mutex.lock();
     self.sync_sender.send(SendMutex(&mutex as _))?;
-    mutex.lock();
+    if !mutex.try_lock_for(Duration::from_secs(30)) {
```
Comment (mmastrac) [inline cli/tools/test/channel.rs:392]: Just in case

**Review by bartlomieju — APPROVED** [review]
Really clever solution, LGTM!

**Pull request merged**